### ISSUE TITLE
UTCDateTime bug for Julian day 360

### ISSUE THREAD

**Issue by tomeracles** [obspy/obspy#2868]
Hello,

I have been using UTCDateTime objects created using year & julian day in the format UTCDateTime('[YYYY][DDD]T'). Unfortunately this seems to work for all days of the year _except_ for Julian day 360, for any year I have tried: the error below appears. I am using ObsPy v. 1.2.2 (conda install obspy) with Python 3.8.10 on MacOS Mojave 10.14.6. Weird!

```
Python 3.8.10 | packaged by conda-forge | (default, May 10 2021, 22:58:09) 
[Clang 11.1.0 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from obspy import UTCDateTime
>>> t = UTCDateTime('2018360T')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Applications/anaconda3/envs/pygmt/lib/python3.8/site-packages/obspy/core/utcdatetime.py", line 424, in __init__
    dt = datetime.datetime.strptime(value, pattern)
  File "/Applications/anaconda3/envs/pygmt/lib/python3.8/_strptime.py", line 568, in _strptime_datetime
    tt, fraction, gmtoff_fraction = _strptime(data_string, format)
  File "/Applications/anaconda3/envs/pygmt/lib/python3.8/_strptime.py", line 352, in _strptime
    raise ValueError("unconverted data remains: %s" %
ValueError: unconverted data remains: 00
```

Apologies if this bug/feature is noted elsewhere, I haven't found it.

Thanks,
Tom

EDIT: For some reason I thought the 'T' was compulsory, however I now realise this was not the case! UTCDateTime('2018360') works fine and seems to be exactly what I want. So I guess this is not a big deal. Still, this is curious, right?

**Comment by calum-chamberlain** [comment]
After some digging this looks like an issue with the `_ISO8601` regular expression. Neither `YYYY360` or `YYYY360T` match the [currently implemented regular expression](https://github.com/obspy/obspy/blob/ae1ed9b30f3643a8dc0e59d4c1ba4b3dbda51739/obspy/core/utcdatetime.py#L27).
The reason `YYYY360` works, but `YYYY360T` doesn't is because the `T` is replaced by a space which is left trailing in the `.split()` line [here](https://github.com/obspy/obspy/blob/ae1ed9b30f3643a8dc0e59d4c1ba4b3dbda51739/obspy/core/utcdatetime.py#L381). A simple fix would be to `.strip()` after the `.replace()` lines above and before `.split()`, however this probably should be caught by the `_ISO8601` regular expression.

Because this isn't documented as a supported option (with the trailing `T`) I'm not going to rush to fix this, but if someone has time to look into that regular expression that would be great!

**Comment by tomeracles** [comment]
Thanks. So is the problem on[ line 34](https://github.com/obspy/obspy/blob/ae1ed9b30f3643a8dc0e59d4c1ba4b3dbda51739/obspy/core/utcdatetime.py#L34), which appears to try to capture all the numbers between 001 and 366:

```      |(00[1-9]|0[1-9]\d|[12]\d{2}|3([0-5]\d|6[1-6]))```

The very last option of [1-6] seems like it should be [0-6] in order to include 360?

**Comment by calum-chamberlain** [comment]
I thought that as well, but didn't test it - I was slightly surprised by other zero-ending julian days matching though, which probably shows my lack of understanding of regular expressions.

**Comment by ThomasLecocq** [comment]
tested with regex101.com:

![image](https://user-images.githubusercontent.com/916937/129673968-c8ad43a0-48c7-444b-a8d4-f50247b97ec4.png)

indeed, changing the last element of the line you mention by ``|6[0-6]`` solves this issue

tom

**Comment by tomeracles** [comment]
NB, at least one of the supported formats specified as an example in the [documentation](https://docs.obspy.org/packages/autogen/obspy.core.utcdatetime.UTCDateTime.html) (under 2: Ordinal date representation) breaks due to this.

```
# Example in docs:
>>> UTCDateTime("2009365T122334.5")            # compact
UTCDateTime(2009, 12, 31, 12, 23, 34, 500000)

# Same example but with julian day 360:
>>> UTCDateTime("2009360T122334.5")            # compact
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Applications/anaconda3/envs/pygmt/lib/python3.8/site-packages/obspy/core/utcdatetime.py", line 424, in __init__
    dt = datetime.datetime.strptime(value, pattern)
  File "/Applications/anaconda3/envs/pygmt/lib/python3.8/_strptime.py", line 568, in _strptime_datetime
    tt, fraction, gmtoff_fraction = _strptime(data_string, format)
  File "/Applications/anaconda3/envs/pygmt/lib/python3.8/_strptime.py", line 352, in _strptime
    raise ValueError("unconverted data remains: %s" %
ValueError: unconverted data remains: 4
```

**Comment by calum-chamberlain** [comment]
Thanks for that point @tomeracles and for checking the regex @ThomasLecocq - I have patched this in #2869. It would be great if you can check this.

**Comment by ThomasLecocq** [comment]
solved with [#130b8694f546f91e40cc7f5b8174b26bd4da2b9d](https://github.com/obspy/obspy/commit/130b8694f546f91e40cc7f5b8174b26bd4da2b9d)
